### PR TITLE
BROOKLYN-597 Remove md5/sha1

### PR DESCRIPTION
**Pull request by geomacy**
Remove MD5 and SHA-1 for [BROOKLYN-597](https://issues.apache.org/jira/projects/BROOKLYN/issues/BROOKLYN-597)

Merge after https://github.com/apache/brooklyn-ui/pull/160

Also simplifies the artifact signing, just hashing the artifacts and signing only the SHA256SUMS file.

Details per JIRA:

Per the recently updated Apache Release Distribution Policy, https://www.apache.org/dev/release-distribution, we should remove the generation and checking of MD5 and SHA-1 checksums from brooklyn-dist/release before we do another release, presumably 1.0.

The relevant wording is 

For every artifact distributed to the public through Apache channels, the PMC

MUST supply a valid OpenPGP-compatible ASCII-armored detached signature file
MUST supply at least one checksum file
SHOULD supply a SHA-256 and/or SHA-512 checksum file
SHOULD NOT supply a MD5 or SHA-1 checksum file (because these are deprecated)
For new releases, PMCs MUST supply SHA-256 and/or SHA-512; and SHOULD NOT supply MD5 or SHA-1. Existing releases do not need to be changed.

### CODE REVIEW

File: release/make-release-artifacts.sh [hunk 1]
```diff
@@ -270,19 +270,21 @@ cp ${src_staging_dir}/brooklyn-dist/deb-packaging/target/apache-brooklyn-${curre
...
-        gpg2 --armor --output ${a}.asc --detach-sig ${a}
-    done
+(
+  cd ${artifact_dir}
+  checksum_file="${artifact_name}-SHA256SUMS"
+
+  for a in *.tar.gz *.zip *.rpm *.deb; do
+      sha256sum -b ${a} >> ${checksum_file}
+  done
+  gpg2 --armor --output ${checksum_file}.asc --detach-sig ${checksum_file}
```
Comment (tbouron) [inline release/make-release-artifacts.sh]: Is this right? My interpretation of the text in the PR description is that the GPG signature should be generated for each artifact, so this should be kept into the for loop. 

Or maybe I misread it?

**Comment by infrastation** [comment]
Not sure if this information helps, but OpenSSL can compute the digests where Linux `coreutils` package is not available, and even print the result in a similar format:
```
$ sha256sum tmp/file.bin 
c00ebfce2fce70f5a56cd6102e82404682bafbc55f2547ef5648bcf9c795a3d0  tmp/file.bin
$ openssl sha256 -r tmp/file.bin 
c00ebfce2fce70f5a56cd6102e82404682bafbc55f2547ef5648bcf9c795a3d0 *tmp/file.bin
$ sha512sum tmp/file.bin 
5aaace31720fb0187660ff0f2eafd265f4ff06fd0722982a33aa3a51bd48238cf918aabdb2db79b15cc2587147374ba1102f6b000061fd6378484066cc9ca06e  tmp/file.bin
$ openssl sha512 -r tmp/file.bin 
5aaace31720fb0187660ff0f2eafd265f4ff06fd0722982a33aa3a51bd48238cf918aabdb2db79b15cc2587147374ba1102f6b000061fd6378484066cc9ca06e *tmp/file.bin
```
There are many other implementations.

File: release/make-release-artifacts.sh (hunk 1 above)
Comment (geomacy) [reply release/make-release-artifacts.sh]: Hm. The policy text quoted in the JIRA doesn't mandate that there has to be a 1:1 correspondence between the artifact and the checksum/signature file, only that there should be a checksum and signature; I was thinking that having one file to do the job for each artifact would be ok.

But you've made me look again at the policy and, lo and behold, it does mandate the 1:1 correspondence in the next paragraph: `The names of signature and checksum files MUST be formed by adding to the name of the artifact the following suffixes:` etc.

So you're right, we need a sum and signature for each artifact independently.

I'll revert that change and push it up again.

**Comment by geomacy** [comment]
Have force pushed it back to 16a3d3f

**Comment by tbouron** [comment]
Thanks @geomacy, LGTM.

Will wait for a "go" from you before merging.

**Comment by tbouron** [comment]
Hi @geomacy, are you happy with this? Let me know what you want to do :)

**Comment by geomacy** [comment]
hi @tbouron, sorry, I have very much had it in my mind to try to get that script working to test this out, but I have not managed to make some time for it. I'm busy pretty much every evening and weekend at the moment, either working or with other events, you know what it's like in the run up to Christmas.

I may have some time to give this a go again on Tuesday or Wednesday evening, and will let you know how I get on.  If you happened to have some time free to try running the release script with this branch, I would be interested to know how you get on. My problem, as mailed on the dev list, was:

I've marked it Do Not Merge because I can't get the make-release-artifacts.sh to work to test this. It keeps failing on brooklyn-ui/ui-modules/utils with
```
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test (default-test) on project brooklyn-software-base: There are test failures.
[ERROR]
[ERROR] Please refer to /private/var/folders/10/7wxypczs3ng9dny8kpnwc9nr0000gn/T/release-working-dir/source/apache-brooklyn-1.0.0-SNAPSHOT-src/brooklyn-server/software/base/target/surefire-reports for the individual test results.
[ERROR] -> [Help 1]
```

**Comment by tbouron** [comment]
Sure, no problem @geomacy, I'll try to build on my local machine with your PR and get back to you with the findings

**Comment by tbouron** [comment]
@geomacy The script worked for me. Just had to do 3 things:
- manually update `make-release-artifacts.sh` to add `-Drpm -Ddeb` to the `mvn` commands (lines 203 & 205) as the deb and rpm packages are now built only when these parameters are given. **This actually should be part of this PR I think.**
- manually update `make-release-artifacts.sh` to use `gpg` instead of `gpg2` (but that is down to my environment)
- update to the latest version of `frontend-maven-plugin`. Version 1.3 was causing a NPE for some reason. I created a PR for it: https://github.com/apache/brooklyn-ui/pull/158 (it was causing a NPE for some reason)

**Comment by geomacy** [comment]
that's terrific @tbouron I will give that a go tonight and include the changes you suggested; will look at ui#158

**Comment by geomacy** [comment]
Well unfortunately I still don't have this going. I caught up to latest masters and added the `rpm` and `deb` flags to the build; this time I didn't get the error in the ui utils directory any more, I guess that is due to your brooklyn-ui#158 @tbouron. But after churning away for over an hour building everything I got to the CLI artifacts stage, which failed as you see:
```
Make CLI artifacts
+ mkdir /var/folders/10/7wxypczs3ng9dny8kpnwc9nr0000gn/T//release-working-dir/bin//apache-brooklyn-1.0.0-SNAPSHOT-client-cli-linux
+ mkdir /var/folders/10/7wxypczs3ng9dny8kpnwc9nr0000gn/T//release-working-dir/bin//apache-brooklyn-1.0.0-SNAPSHOT-client-cli-windows
+ mkdir /var/folders/10/7wxypczs3ng9dny8kpnwc9nr0000gn/T//release-working-dir/bin//apache-brooklyn-1.0.0-SNAPSHOT-client-cli-macosx
+ cp /var/folders/10/7wxypczs3ng9dny8kpnwc9nr0000gn/T//release-working-dir/source/apache-brooklyn-1.0.0-SNAPSHOT-src/brooklyn-client/cli/target/bin/linux.386/br /var/folders/10/7wxypczs3ng9dny8kpnwc9nr0000gn/T//release-working-dir/bin//apache-brooklyn-1.0.0-SNAPSHOT-client-cli-linux
cp: /var/folders/10/7wxypczs3ng9dny8kpnwc9nr0000gn/T//release-working-dir/source/apache-brooklyn-1.0.0-SNAPSHOT-src/brooklyn-client/cli/target/bin/linux.386/br: No such file or directory
```

My build command was `./brooklyn-dist/release/make-release-artifacts.sh  -r1 -n` issued from the Brooklyn root dir.

I don't know what's up here and it's a bit late to start investigating that - I have pushed the addition flags commit.

I "tested" the signing changes by running just that chunk of the script, and it succeeds to the extent that it signs the artifacts that it finds in the directory as it should, but in my case it fails because there is no .rpm file in the directory - due to the real build failing as above.

@tbouron I'm fairly confident in the correctness of the changes but can't demonstrate them yet. Also I won't be able to get to look at this stuff again at least until Thursday. Happy to do that but if you want to do anything with it yourself and/or merge it on the basis of your own tests, feel free. I've invited you as a collaborator on my repo so you can update my branch if you like.

**Comment by geomacy** [comment]
I have added a few more commits, which I found were required, along with https://github.com/apache/brooklyn-ui/pull/160. This allows the script to succeed and builds 
```
The release is done - here is what has been created:
apache-brooklyn-1.0.0-rc1-1.noarch.rpm				apache-brooklyn-1.0.0-rc1-client-cli-macosx.zip
apache-brooklyn-1.0.0-rc1-1.noarch.rpm.asc			apache-brooklyn-1.0.0-rc1-client-cli-macosx.zip.asc
apache-brooklyn-1.0.0-rc1-1.noarch.rpm.sha256			apache-brooklyn-1.0.0-rc1-client-cli-macosx.zip.sha256
apache-brooklyn-1.0.0-rc1-bin.tar.gz				apache-brooklyn-1.0.0-rc1-client-cli-windows.tar.gz
apache-brooklyn-1.0.0-rc1-bin.tar.gz.asc			apache-brooklyn-1.0.0-rc1-client-cli-windows.tar.gz.asc
apache-brooklyn-1.0.0-rc1-bin.tar.gz.sha256			apache-brooklyn-1.0.0-rc1-client-cli-windows.tar.gz.sha256
apache-brooklyn-1.0.0-rc1-bin.zip				apache-brooklyn-1.0.0-rc1-client-cli-windows.zip
apache-brooklyn-1.0.0-rc1-bin.zip.asc				apache-brooklyn-1.0.0-rc1-client-cli-windows.zip.asc
apache-brooklyn-1.0.0-rc1-bin.zip.sha256			apache-brooklyn-1.0.0-rc1-client-cli-windows.zip.sha256
apache-brooklyn-1.0.0-rc1-classic.tar.gz			apache-brooklyn-1.0.0-rc1-src.tar.gz
apache-brooklyn-1.0.0-rc1-classic.tar.gz.asc			apache-brooklyn-1.0.0-rc1-src.tar.gz.asc
apache-brooklyn-1.0.0-rc1-classic.tar.gz.sha256			apache-brooklyn-1.0.0-rc1-src.tar.gz.sha256
apache-brooklyn-1.0.0-rc1-classic.zip				apache-brooklyn-1.0.0-rc1-src.zip
apache-brooklyn-1.0.0-rc1-classic.zip.asc			apache-brooklyn-1.0.0-rc1-src.zip.asc
apache-brooklyn-1.0.0-rc1-classic.zip.sha256			apache-brooklyn-1.0.0-rc1-src.zip.sha256
apache-brooklyn-1.0.0-rc1-client-cli-linux.tar.gz		apache-brooklyn-1.0.0-rc1-vagrant.tar.gz
apache-brooklyn-1.0.0-rc1-client-cli-linux.tar.gz.asc		apache-brooklyn-1.0.0-rc1-vagrant.tar.gz.asc
apache-brooklyn-1.0.0-rc1-client-cli-linux.tar.gz.sha256	apache-brooklyn-1.0.0-rc1-vagrant.tar.gz.sha256
apache-brooklyn-1.0.0-rc1-client-cli-linux.zip			apache-brooklyn-1.0.0-rc1-vagrant.zip
apache-brooklyn-1.0.0-rc1-client-cli-linux.zip.asc		apache-brooklyn-1.0.0-rc1-vagrant.zip.asc
apache-brooklyn-1.0.0-rc1-client-cli-linux.zip.sha256		apache-brooklyn-1.0.0-rc1-vagrant.zip.sha256
apache-brooklyn-1.0.0-rc1-client-cli-macosx.tar.gz		apache-brooklyn-1.0.0-rc1.deb
apache-brooklyn-1.0.0-rc1-client-cli-macosx.tar.gz.asc		apache-brooklyn-1.0.0-rc1.deb.asc
apache-brooklyn-1.0.0-rc1-client-cli-macosx.tar.gz.sha256	apache-brooklyn-1.0.0-rc1.deb.sha256
You can find these files in: /Users/geoff/workspaces/github.com/apache/brooklyn/brooklyn-dist/release/apache-brooklyn-1.0.0-rc1
```

**Comment by tbouron** [comment]
Tested with the last commits, it now works, great job @geomacy 👏 

Although as you said, it requires https://github.com/apache/brooklyn-ui/pull/160 to be merged first

**Comment by ahgittin** [comment]
merging -- great stuff!

**Comment by geomacy** [comment]
Must update the docs for this next!

**Pull request merged**